app/src/main/java/woowacourse/shopping/presentation/detail/DetailActivity.kt

```diff
@@ -0,0 +1,77 @@
+package woowacourse.shopping.presentation.detail
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.view.Menu
+import android.view.MenuItem
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import androidx.databinding.DataBindingUtil
+import androidx.lifecycle.ViewModelProvider
+import woowacourse.shopping.R
+import woowacourse.shopping.data.datasource.DefaultCart
+import woowacourse.shopping.data.datasource.DefaultProducts
+import woowacourse.shopping.data.repository.CartRepositoryImpl
+import woowacourse.shopping.data.repository.ProductRepositoryImpl
+import woowacourse.shopping.databinding.ActivityDetailBinding
+
+class DetailActivity : AppCompatActivity(), AddCompleteListener {
+    private val binding: ActivityDetailBinding by lazy {
+        DataBindingUtil.setContentView(this, R.layout.activity_detail)
+    }
+    private val viewModel: DetailViewModel by lazy {
+        ViewModelProvider(
+            this,
+            DetailViewModelFactory(ProductRepositoryImpl(DefaultProducts), CartRepositoryImpl(DefaultCart)),
+        )[DetailViewModel::class.java]
+    }
```

Comment:
기존의 구현은 화면마다 매번 implementation 객체를 생성해야 하는 비효율이 있을 것이라고 생각합니다. 화면 별로 Repository 차원에서 관리해야 하는 데이터가 존재하지 않기 때문에, Application 실행 시부터 끝까지 전역적으로 관리하고 필요한 곳에서 꺼내어 사용하는 것이 적절하다는 생각이 들었습니다. 그에 따라 다음과 같이 변경을 진행하였습니다.
```
class ShoppingApplication : Application() {
    val productRepository: ProductRepository by lazy {
        ProductRepositoryImpl(DefaultProducts)
    }
    val cartRepository: CartRepository by lazy {
        CartRepositoryImpl(DefaultCart)
    }
}
```

다만 여기서 궁금한 점이 있었습니다. 과연 companion object 등을 통한 캐싱으로 구현하는 것과 커다란 차이가 있는 것인지에 관한 것이었습니다. 실제로 저는 지난 미션에서 데이터베이스에 대한 캐싱을 진행하기 위해 companion object를 활용하곤 했습니다.
```
abstract class MovieDatabase : RoomDatabase() {
    companion object {
        // 캐싱
        @Volatile
        private var synchronizedInstance: MovieDatabase? = null

        fun getDatabase(context: Context): MovieDatabase {
            return synchronizedInstance ?: synchronized(this) {
                val instance =
                    Room.databaseBuilder(/* ... */)
                        .build()
                synchronizedInstance = instance
                instance
            }
        }
    }
}
```
코드를 토대로 생각해보았을 때, 제시해주신 코드랩 대로의 데이터베이스 초기화와 저의 구현대로의 데이터베이스 초기화는 성능 상 차이가 없다고 생각하였습니다. (application 단에서의 선언은 by lazy를 통하여 지연 초기화를 진행하고, companion object를 사용한 구현 역시 역시 함수를 통해 생성하기 때문이라고 생각했습니다.) 

따라서 이러한 구현의 차이는 혹시 취향차이로 생각해볼 수 있는 것인지 질문드리고 싶습니다...!

Reply:
둘은 역할이 다릅니다.
제시해주신 코드의 `MovieDatabase` 내 companion object 의 `getDatabase()` 함수는 java의 static 한 범위로 어디서든 접근이 가능하여 getDataBase 함수 호출을 통한 인스턴스 초기화를 가능하도록 제공해주고 있습니다. 즉, 호출하기 전까지는 인스턴스 초기화가 되지 않는것이죠.

이를 이용해 Application 에서는 lazy 키워드를 통해 dataBase 인스턴스를 호출할 때 초기화를 하도록 해주는 키워드입니다. 만약, dataBase 인스턴스를 Application 내 companion object 로 선언한다면 Application 클래스가 만들어질 때 함께 생성이 되겠죠.
때문에 위 링크에서도 아래와 같은 설명이 있습니다.

> Because these objects should only be created when they're first needed, rather than at app startup, you're using Kotlin's property delegation: [by lazy](https://kotlinlang.org/docs/reference/delegated-properties.html#lazy).